### PR TITLE
Promote PodDisruptionBudget to GA

### PR DESCRIPTION
**Pull request by mortent**
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:
Promotes PodDisruptionBudget type to v1

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
The `PodDisruptionBudget` API have been promoted to `policy/v1`.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
- KEP: https://github.com/kubernetes/enhancements/pull/904

```docs

```

### CODE REVIEW

**Review by liggitt — COMMENTED** [review]
> @liggitt PDBs already have e2e tests in `test/e2e/apps/disruption.go`, but these are obviously using the v1beta1 API. Should I have those over to use the v1 API or will that create problems with the skew tests?

Tests in master can use v1 APIs introduced in master. Since this is promoting with no changes, switching the existing e2e's to v1 is reasonable.

File: staging/src/k8s.io/api/policy/v1/types.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+// +genclient
+// +genclient:noVerbs
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Eviction evicts a pod from its node subject to certain policies and safety constraints.
+// This is a subresource of Pod.  A request to cause such an eviction is
+// created by POSTing to .../pods/<pod name>/evictions.
+type Eviction struct {
```
[the remark this answers is not in the document]
Comment (lavalamp) [reply staging/src/k8s.io/api/policy/v1/types.go]: I don't know if we've ever version bumped a subresource yet. The intention was to use headers to communicate types the client can understand (for GETable resources), which this isn't. But I think we can't add the type unless it actually can be POSTed to the subresource. I don't see the code to do that in this PR, but maybe I missed it?

File: api/api-rules/violation_exceptions.list [hunk 2]
```diff
@@ -270,6 +270,7 @@ API rule violation: list_type_missing,k8s.io/api/networking/v1beta1,IngressSpec,
 API rule violation: list_type_missing,k8s.io/api/networking/v1beta1,IngressTLS,Hosts
 API rule violation: list_type_missing,k8s.io/api/node/v1alpha1,RuntimeClassList,Items
 API rule violation: list_type_missing,k8s.io/api/node/v1beta1,RuntimeClassList,Items
+API rule violation: list_type_missing,k8s.io/api/policy/v1,PodDisruptionBudgetList,Items
```
Comment (lavalamp) [inline api/api-rules/violation_exceptions.list]: Please fix items rather than add to this list.

File: staging/src/k8s.io/api/policy/v1/types.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	// and the value is the time when the API server processed the eviction request. If
+	// the deletion didn't occur and a pod is still there it will be removed from
+	// the list automatically by PodDisruptionBudget controller after some time.
+	// If everything goes smooth this map should be empty for the most of the time.
+	// Large number of entries in the map may indicate problems with pod deletions.
+	// +optional
+	DisruptedPods map[string]metav1.Time `json:"disruptedPods,omitempty" protobuf:"bytes,2,rep,name=disruptedPods"`
+
+	// podDisruptionsAllowed specifies the number of pod disruptions that are currently allowed.
+	PodDisruptionsAllowed int32 `json:"disruptionsAllowed" protobuf:"varint,3,opt,name=disruptionsAllowed"`
```
Comment (lavalamp) [inline staging/src/k8s.io/api/policy/v1/types.go]: json tag doesn't match the go name, fix the go name probably.

File: api/api-rules/violation_exceptions.list [hunk 4]
```diff
@@ -516,6 +517,7 @@ API rule violation: names_match,k8s.io/api/core/v1,RBDVolumeSource,RBDPool
 API rule violation: names_match,k8s.io/api/core/v1,RBDVolumeSource,RadosUser
 API rule violation: names_match,k8s.io/api/core/v1,VolumeSource,CephFS
 API rule violation: names_match,k8s.io/api/core/v1,VolumeSource,StorageOS
+API rule violation: names_match,k8s.io/api/policy/v1,PodDisruptionBudgetStatus,PodDisruptionsAllowed
```
Comment (liggitt) [inline api/api-rules/violation_exceptions.list]: this is the json tag mismatch violation... fixing the field name will fix this as well. unless you want to do custom conversions (pro-tip, you don't), you should change the field name in the internal and v1beta1 types as well

File: staging/src/k8s.io/api/policy/v1/types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/api/policy/v1/types.go]: > I don't know if we've ever version bumped a subresource yet. The intention was to use headers to communicate types the client can understand (for GETable resources), which this isn't. But I think we can't add the type unless it actually can be POSTed to the subresource. I don't see the code to do that in this PR, but maybe I missed it?

we haven't, and currently clients certainly aren't expecting the version they can submit to a given endpoint to change out from under them. I agree we shouldn't make the v1 type for Eviction until we can use/test it

[the remark this answers is not in the document]
Comment (mortent) [reply staging/src/k8s.io/api/policy/v1/types.go]: I have removed the v1 type for Eviction.

**Comment by mortent** [comment]
I think all the issues have been addressed now.

File: pkg/registry/policy/poddisruptionbudget/strategy.go [hunk 5]
```diff
@@ -109,7 +111,8 @@ func (podDisruptionBudgetStatusStrategy) PrepareForUpdate(ctx context.Context, o
 
 // ValidateUpdate is the default update validation for an end user updating status
 func (podDisruptionBudgetStatusStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
-	// TODO: Validate status updates.
```
Comment (liggitt) [inline pkg/registry/policy/poddisruptionbudget/strategy.go]: flip this around to return early for policy/v1beta1 for backwards compatibility, but to validate in all other cases. that ensures we continue validating if future versions are introduced (v2alpha1, etc)

Comment (liggitt) [reply pkg/registry/policy/poddisruptionbudget/strategy.go]: also, add a unit test that makes sure this does not call validation for `policy/v1beta1`, does call it for `policy/v1` and `policy/random-version`

Comment (mortent) [reply pkg/registry/policy/poddisruptionbudget/strategy.go]: I have fixed the logic in the `ValidateUpdate` method and added a unit test.

**Comment by liggitt** [comment]
xref https://github.com/kubernetes/kubernetes/issues/77383

**Comment by mortent** [comment]
@liggitt I have looked at #77383 and it does not affect the effort to move PDBs to GA.

**Comment by liggitt** [comment]
> @liggitt I have looked at #77383 and it does not affect the effort to move PDBs to GA.

you mention providing better feedback in that scenario... does the API provide a mechanism for doing that (e.g. status conditions)?

**Comment by liggitt** [comment]
Is this considered an optional or required API? If required, this will need e2e tests exercising the v1 APIs being introduced that can be promoted to conformance after soaking. See https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-sig-leads/3lrI2lCzxPk/6hUhIK3HDwAJ for details

See https://apisnoop.cncf.io/?bucket=apisnoop%2Fdev%2Fci-kubernetes-e2e-gci-gce%2F1183553991464718336&zoomed=category-beta-policy for current e2e coverage of the beta APIs (though it's unclear how accurate that dashboard is, see https://github.com/cncf/apisnoop/issues/271)

File: pkg/registry/policy/poddisruptionbudget/strategy.go (hunk 5 above)
Comment (liggitt) [reply pkg/registry/policy/poddisruptionbudget/strategy.go]: also compare `policyapiv1beta1.SchemeGroupVersion.Group == requestInfo.APIGroup`

**Comment by mortent** [comment]
I need to look at ways to improve feedback. Status conditions would be my guess on how we might do it. PDBs currently does not have status conditions, so we need to figure out which conditions make sense for PDBs and try to keep them consistent with the conditions in other resources. Would it be a problem to add this after GA?

We already have e2e tests for PDBs, although they are under apps rather than policy: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/disruption.go

**Comment by liggitt** [comment]
> We already have e2e tests for PDBs, although they are under apps rather than policy: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/disruption.go

I see 12 API calls for PDBs under test/e2e:
```
$ grep -R .PodDisruptionBudgets test/e2e
test/e2e/framework/service/jig.go:	newPdb, err := j.Client.PolicyV1beta1().PodDisruptionBudgets(j.Namespace).Create(pdb)
test/e2e/framework/service/jig.go:		pdb, err := j.Client.PolicyV1beta1().PodDisruptionBudgets(j.Namespace).Get(j.Name, metav1.GetOptions{})
test/e2e/autoscaling/cluster_size_autoscaling.go:	_, err = f.ClientSet.PolicyV1beta1().PodDisruptionBudgets(namespace).Create(pdb)
test/e2e/autoscaling/cluster_size_autoscaling.go:		f.ClientSet.PolicyV1beta1().PodDisruptionBudgets(namespace).Delete(pdb.Name, &metav1.DeleteOptions{})
test/e2e/autoscaling/cluster_size_autoscaling.go:	ginkgo.By("Create PodDisruptionBudgets for kube-system components, so they can be migrated if required")
test/e2e/autoscaling/cluster_size_autoscaling.go:			err := f.ClientSet.PolicyV1beta1().PodDisruptionBudgets("kube-system").Delete(newPdbName, &metav1.DeleteOptions{})
test/e2e/autoscaling/cluster_size_autoscaling.go:		_, err := f.ClientSet.PolicyV1beta1().PodDisruptionBudgets("kube-system").Create(pdb)
test/e2e/apps/disruption.go:			pdb, err := cs.PolicyV1beta1().PodDisruptionBudgets(ns).Get("foo", metav1.GetOptions{})
test/e2e/apps/disruption.go:	_, err := cs.PolicyV1beta1().PodDisruptionBudgets(ns).Create(&pdb)
test/e2e/apps/disruption.go:	_, err := cs.PolicyV1beta1().PodDisruptionBudgets(ns).Create(&pdb)
test/e2e/apps/disruption.go:		old, err := cs.PolicyV1beta1().PodDisruptionBudgets(ns).Get("foo", metav1.GetOptions{})
test/e2e/apps/disruption.go:		if _, err := cs.PolicyV1beta1().PodDisruptionBudgets(ns).Update(old); err != nil {
test/e2e/apps/disruption.go:		pdb, err := cs.PolicyV1beta1().PodDisruptionBudgets(ns).Get("foo", metav1.GetOptions{})
```

This is only updating two of those calls to use the v1 API.

@johnbelamaric what would you envision as the recommended path for promoting an API while maintaining test coverage, per https://groups.google.com/forum/#!msg/kubernetes-sig-leads/3lrI2lCzxPk/6hUhIK3HDwAJ ?

**Comment by johnbelamaric** [comment]
The tests should be against v1. The fact that the beta APIs are no longer tested affects overall coverage but does not affect conformance coverage since those aren't subject to conformance.

We do have the problem though that if we promote the API to GA prior to promoting the tests to conformance, then we see a drop in conformance coverage during the gap. If we have a blocking check for that a la https://github.com/kubernetes/enhancements/pull/1306, we have a problem.

The cleanest solution would be that the tests are promoted to conformance at the same time the API is promoted to GA. I am hesitant to insist on that, though, as it makes the reviews even more painful. It requires review of all those tests to validate they meet the conformance constraints. I am assuming since they have been around a while they are already known to not be flaky.

This problem should be discussed on https://github.com/kubernetes/enhancements/pull/1306 and solved there; I don't think we should hold up this PR for it. As a start though, I would like to see a separate PR that does the promotion for these tests, targeted at the same milestone, and referenced from this PR.

**Comment by markyjackson-taulia** [comment]
Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

**Comment by mortent** [comment]
@liggitt I had missed some of the e2e-tests that used PDBs. I have updated the PR.

@johnbelamaric I will look at creating a PR to promote PDB e2e tests to conformance tests.

**Comment by mortent** [comment]
I have created a PR to promote e2e tests for PDBs to conformance: #84740

**Comment by mortent** [comment]
@liggitt @lavalamp I think this should be ready now. Promotion of e2e-tests to conformance is done in #84740. I would like to look into adding conditions separately from this as it will require a deeper look at the controller to identify the correct conditions.

@markyjackson-taulia I hope to get this into 1.17.

**Comment by markyjackson-taulia** [comment]
@liggitt @lavalamp @mortent, this has conflicts

File: staging/src/k8s.io/api/policy/v1beta1/types.go [hunk 6]
```diff
@@ -67,7 +67,7 @@ type PodDisruptionBudgetStatus struct {
 	DisruptedPods map[string]metav1.Time `json:"disruptedPods,omitempty" protobuf:"bytes,2,rep,name=disruptedPods"`
 
 	// Number of pod disruptions that are currently allowed.
-	PodDisruptionsAllowed int32 `json:"disruptionsAllowed" protobuf:"varint,3,opt,name=disruptionsAllowed"`
+	DisruptionsAllowed int32 `json:"disruptionsAllowed" protobuf:"varint,3,opt,name=disruptionsAllowed"`
```
Comment (lavalamp) [inline staging/src/k8s.io/api/policy/v1beta1/types.go]: FWIW this rename made this PR a fair amount bigger than it needed to be, maybe next time do this and the promotion in separate steps? This change by itself would have been mechanical and super easy to review.

Comment (mortent) [reply staging/src/k8s.io/api/policy/v1beta1/types.go]: Yeah, next time I will definitely do these kind of changes in separate PRs.

**Comment by mortent** [comment]
The conflicts has been resolved, so this is ready for review.

**Comment by markyjackson-taulia** [comment]
@liggitt @lavalamp @mortent 24 hours till code freeze, we still going to make v1.17?

**Comment by liggitt** [comment]
This PR is in good shape, but having swept open issues/PRs around PDBs (organized in https://github.com/orgs/kubernetes/projects/36), there still looks to be work needed before we call it GA. Let's coordinate in the linked project board to work through those issues and add test coverage so this can proceed in 1.18.

/milestone v1.18

**Comment by annajung** [comment]
Hello @mortent   👋 I'm one of the Bug Triage shadows for the 1.18 release. Since this PR is tagged for 1.18, I'd like to check its status. Is this PR intended for milestone 1.18? 

Friendly reminder: Code freeze is scheduled for March 5th.

**Comment by annajung** [comment]
Hello @mortent @liggitt Checking with you again. Is this PR intended for milestone 1.18?

**Comment by annajung** [comment]
Hey @mortent @liggitt Code freeze is coming up in a few days and wanted to check in with you to see if this will be completed for the release?

**Comment by mortent** [comment]
@annajung There are still open issues. This will not make it for 1.18.

**Comment by annajung** [comment]
@mortent Thank you for the update. I will defer the milestone to v.1.19.
/milestone v1.19

**Comment by bai** [comment]
👋 Hello from Bug Triage team! This PR has not been updated for a while, so I'd like to check its status. The code freeze is starting June 25th (about 4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.19, is it still planned for this release?

**Comment by seh** [comment]
See #95083 for a matter we could address before graduating.

**Pull request closed without merging**